File: python/algos/levenshtein_distance.py
def levenshtein_distance(s1, s2):
    l1, l2 = len(s1), len(s2)
    prev_row, curr_row = [i for i in range(l2+1)], [0]*(l2+1)
    for i in range(1, l1+1):
        curr_row[0] = i
        for j in range(1, l2+1):
            if s1[i-1] == s2[j-1]:
                curr_row[j] = prev_row[j-1]
            else:
                curr_row[j] = 1+min(curr_row[j-1], prev_row[j], prev_row[j-1])
        prev_row = curr_row.copy()
    return curr_row[l2]

File: algos/levenshteinDistance.cpp
#include "levenshteinDistance.h"

int LevenshteinDistance::score(std::string s1, std::string s2) {
    const int len1 = s1.length();
    const int len2 = s2.length();
    int prevRow[len2 + 1];
    int currRow[len2 + 1];
    for (int i = 0; i < len2 + 1; i++) {
        prevRow[i] = i;
        currRow[i] = 0;
    }
    for (int i = 1; i < len1 + 1; i++) {
        currRow[0] = i;
        for (int j = 1; j < len2 + 1; j++) {
            if (s1[i-1] == s2[j-1]) {
                currRow[j] = prevRow[j-1];
            } else {
                currRow[j] = 1 + std::min({currRow[j-1], prevRow[j], prevRow[j-1]});
            }
        }
        for (int j = 0; j < len2; j++) {
            prevRow[j] = currRow[j];
        }
    }
    return currRow[len2];
}

